src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResultBuilder.cs

```diff
@@ -122,31 +158,65 @@ internal void ApplySuppressionsAndStoreAnalysisResult(AnalysisScope analysisScop
                     }
                 }
             }
+
+            static SyntaxTree? getSourceTree(Diagnostic diagnostic)
+                => diagnostic.Location.SourceTree;
+
+            AdditionalText? getAdditionalTextKey(Diagnostic diagnostic)
+            {
+                // Fetch the first additional file that matches diagnostic location.
+                if (diagnostic.Location is ExternalFileLocation externalFileLocation)
+                {
+                    var path = externalFileLocation.FilePath;
+                    if (_pathToAdditionalTextMap.TryGetValue(externalFileLocation.FilePath, out var additionalTexts))
+                    {
+                        if (additionalTexts.Count == 1)
+                        {
+                            return additionalTexts[0];
+                        }
+
+                        foreach (var additionalText in additionalTexts)
+                        {
+                            if (analysisScope.NonSourceFiles.Contains(additionalText))
+                            {
+                                return additionalText;
+                            }
+                        }
```

Comment:
Feel like there is a missing `throw` statement or assert here. Given that we unconditionally return in the case of a single file (don't do the `NonSourceFiles.Contains` check) there appears to be an implicit assumption the data is correct. Basically if there is one file it must be the correct one so use it. 

In the case of multiple files though we do extra checking. But shouldn't that extra checking always match a value? It feels like if we get to the end of the `foreach` and have not matched a value then something is wrong here.

Reply:
Thanks, I have fixed the code by removing the special case for single file case.